python/pyarrow/_compute.pyx

```diff
@@ -2568,31 +2580,53 @@ cdef inline CFunctionDoc _make_function_doc(dict func_doc) except *:
     return f_doc
 
 
-cdef object box_scalar_udf_context(const CScalarUdfContext& c_context):
-    cdef ScalarUdfContext context = ScalarUdfContext.__new__(ScalarUdfContext)
+cdef object box_udf_context(const CUdfContext& c_context):
+    cdef UdfContext context = UdfContext.__new__(UdfContext)
     context.init(c_context)
     return context
 
 
-cdef _scalar_udf_callback(user_function, const CScalarUdfContext& c_context, inputs):
+cdef _udf_callback(user_function, const CUdfContext& c_context, inputs):
     """
-    Helper callback function used to wrap the ScalarUdfContext from Python to C++
+    Helper callback function used to wrap the UdfContext from Python to C++
     execution.
     """
-    context = box_scalar_udf_context(c_context)
+    context = box_udf_context(c_context)
     return user_function(context, *inputs)
 
 
-def _get_scalar_udf_context(memory_pool, batch_length):
-    cdef CScalarUdfContext c_context
+def _get_udf_context(memory_pool, batch_length):
+    cdef CUdfContext c_context
     c_context.pool = maybe_unbox_memory_pool(memory_pool)
     c_context.batch_length = batch_length
-    context = box_scalar_udf_context(c_context)
+    context = box_udf_context(c_context)
     return context
 
 
-def register_scalar_function(func, function_name, function_doc, in_types,
-                             out_type):
+ctypedef CStatus (*CRegisterUdf)(PyObject* function, function[CallbackUdf] wrapper,
+                                 const CUdfOptions& options, CFunctionRegistry* registry)
+
+cdef class RegisterUdf(_Weakrefable):
+    cdef CRegisterUdf register_func
+
+    cdef void init(self, const CRegisterUdf register_func):
+        self.register_func = register_func
+
+
+cdef get_register_scalar_function():
+    cdef RegisterUdf reg = RegisterUdf.__new__(RegisterUdf)
+    reg.register_func = RegisterScalarFunction
+    return reg
+
+
+cdef get_register_tabular_function():
+    cdef RegisterUdf reg = RegisterUdf.__new__(RegisterUdf)
+    reg.register_func = RegisterTabularFunction
+    return reg
+
```

Comment:
appreciate your effort to make this more generic, but with the other UDFs, we have a complex strucutre for the callbacks. In that case this generalization won't be that useful as far as I feel. For the moment, shall we keep them separate? There would be a sort of `DRY` violated here, but once we generalized this API after our experimental version, we should be able to streamline these. 

cc @westonpace

Reply:
You mean revert to the original `register_scalar_udf` code, then duplicate it and adapt to `register_tabular_udf`? The disadvantage of code repetition is clear, but what is the advantage? The remaining PR would still need to resolve the conflict, since you say the callback structure changed in that PR, only this time the conflict would not be caught by source control tools, because the `register_scalar_udf` code would not have changed.

Reply:
The proposed common design for scalar and tabular function registration requires a parameter to distinguish between the two. I considered 3 options to do so:
1. Using a Boolean parameter, such as `is_tabular`. This seemed the least elegant to me, and is definitely the least extensible.
2. Using and enum parameter, which would currently have two values for scalar and tabular. This is better but seemed undesirable because it is not clear what this enum is. Is it the function's kind? the function output's kind? something else? See also [this post](https://github.com/apache/arrow/pull/14682/#discussion_r1050597587).
3. Using an interface parameter. This seemed elegant and extensible to me. The interface just describes how the function is registered, and is easy enough to implement. At least for the time being, the intention is for this interface to be internal.

Reply:
My preference would be to keep this and, if we have something better when we add aggregate udfs, we can switch to that then.  That being said, I'm not 100% certain I follow the arguments here.